usaspending_api/data_load/data_load_helpers.py

```diff
@@ -0,0 +1,32 @@
+import datetime
+
+subtier_agency_list = {}  # global variable, populated by fpds_loader
+
+
+def capitalize_if_string(val):
+    if isinstance(val, str):
+        return val.upper()
+    return val
+
+
+def false_if_null(val):
+    if val is None:
+        return False
+    return val
+
+
+# TODO: replace this with cursor.morgify() in some way that doesn't need a live connection passed around everywhere
+def format_value_for_sql(val):
```

Comment:
Yes. As far as I could find the only option is cursor.morgify(), which requires an open connection to do any formatting. Right now I keep the connection closed during the conversion from Broker's schema to ours to not throttle connections, and I really don't like passing around a cursor this deep into helper methods. I left the TODO there because this is a concern

Reply:
It might be too much of a lift since I'm still learning this script. But there are psycopg2 helper functions which should asssit: from psycopg2.sql import Identifier, Literal, SQL, ...

Also, when executing SQL it is possible to separate the SQL and the values. The main benefit is to protect against SQL injection (which isn't a huge concern here) it should also help to massage the data into DB appropriate types

Reply:
I recommend looking into psycopg Composable objects.  I use them all the time.  You don't need to worry about injection or formatting or anything.  It handles it all.

If interested, some examples can be found hur: https://github.com/fedspendingtransparency/usaspending-api/blob/dev/usaspending_api/awards/v2/views/accounts.py

The key concepts are to wrap your sql in a SQL object, wrap values in Literal, and wrap object names (tables, columns, etc) in Identifier.

Reply:
Looks deceptively simple. I'm going to try blindly applying it and see what crashes

Reply:
I've now tried it out, and while it does work (I think), it more than halves performance. Since there's only a limited number of values that we might be dealing with, I would advise KEEPING the less safe option in this version of the PR

Reply:
One caveat, though ... for some reason they're not compatible with the `django-debug-toolbar` so if you're using Django to run your queries you need to convert them to actual raw SQL (look for a function called `convert_composable_query_to_string`).  It looks like you're using psycopg to run your queries, so this shouldn't apply to you.  Just a heads up.

Reply:
The original objective of this branch was to NOT use Django